sky/compositor/raster_cache.cc

```diff
@@ -0,0 +1,86 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "sky/compositor/raster_cache.h"
+
+#include "sky/compositor/paint_context.h"
+#include "base/logging.h"
+#include "third_party/skia/include/core/SkImage.h"
+
+namespace sky {
+namespace compositor {
+
+static const int kRasterThreshold = 3;
+
+static bool isWorthRasterizing(SkPicture* picture) {
+  // TODO(abarth): We should find a better heuristic here that lets us avoid
+  // wasting memory on trivial layers that are easy to re-rasterize every frame.
+  return picture->approximateOpCount() > 10 || picture->hasText();
+}
+
+RasterCache::RasterCache() {
+}
+
+RasterCache::~RasterCache() {
+}
+
+RasterCache::Entry::Entry() {
+  physical_size.setEmpty();
+}
+
+RasterCache::Entry::~Entry() {
+}
+
+RefPtr<SkImage> RasterCache::GetImage(
+    SkPicture* picture, const SkISize& physical_size) {
+  if (physical_size.isEmpty())
+    return nullptr;
+
+  Entry& entry = cache_[picture->uniqueID()];
+
+  const bool size_matched = entry.physical_size == physical_size;
+
+  entry.used = true;
+  entry.physical_size = physical_size;
+
+  if (!size_matched) {
+    entry.access_count = 1;
+    entry.image = nullptr;
+    return nullptr;
+  }
+
+  entry.access_count++;
+
+  if (entry.access_count >= kRasterThreshold) {
+    // Saturate at the threshhold.
+    entry.access_count = kRasterThreshold;
+
+    if (!entry.image && isWorthRasterizing(picture)) {
+      entry.image = adoptRef(SkImage::NewFromPicture(picture, physical_size,
+                                                     nullptr, nullptr));
+    } else if (entry.image) {
+      entry.reused_image = true;
+    }
+  }
+
+  return entry.image;
+}
+
+void RasterCache::Sweep() {
```

Comment:
rename this to "SweepAfterFrame" or something or document that it's intended to be used at the end of a frame (and is not idempotent).